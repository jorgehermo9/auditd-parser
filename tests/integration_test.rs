use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
