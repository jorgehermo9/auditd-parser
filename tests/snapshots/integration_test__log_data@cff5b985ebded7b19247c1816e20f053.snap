---
source: tests/integration_test.rs
expression: record
info: "type=EXECVE msg=audit(1655465398.534:25618): argc=48 a0=\"/bin/sh\" a1=\"-efu\" a2=\"/usr/bin/ld\" a3=\"-plugin\" a4=\"/usr/libexec/gcc/aarch64-alt-linux/8/liblto_plugin.so\" a5=\"-plugin-opt=/usr/libexec/gcc/aarch64-alt-linux/8/lto-wrapper\" a6=\"-plugin-opt=-fresolution=/usr/src/tmp/cchyHiZN.res\" a7=\"-plugin-opt=-pass-through=-lgcc\" a8=\"-plugin-opt=-pass-through=-lgcc_s\" a9=\"-plugin-opt=-pass-through=-lc\" a10=\"-plugin-opt=-pass-through=-lgcc\" a11=\"-plugin-opt=-pass-through=-lgcc_s\" a12=\"--build-id\" a13=\"--no-add-needed\" a14=\"--eh-frame-hdr\" a15=\"--hash-style=gnu\" a16=\"--as-needed\" a17=\"-shared\" a18=\"-X\" a19=\"-EL\" a20=\"-maarch64linux\" a21=\"-o\" a22=\"ztest105133.so\" a23=\"/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64/crti.o\" a24=\"/usr/lib64/gcc/aarch64-alt-linux/8/crtbeginS.o\" a25=\"-L/usr/lib64/gcc/aarch64-alt-linux/8\" a26=\"-L/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64\" a27=\"-L/lib/../lib64\" a28=\"-L/usr/lib/../lib64\" a29=\"-L/usr/lib64/gcc/aarch64-alt-linux/8/../../..\" a30=\"-soname\" a31=\"libz.so.1\" a32=\"--version-script\" a33=\"zlib.map\" a34=\"ztest105133.o\" a35=\"-lgcc\" a36=\"--push-state\" a37=\"--as-needed\" a38=\"-lgcc_s\" a39=\"--pop-state\" a40=\"-lc\" a41=\"-lgcc\" a42=\"--push-state\" a43=\"--as-needed\" a44=\"-lgcc_s\" a45=\"--pop-state\" a46=\"/usr/lib64/gcc/aarch64-alt-linux/8/crtendS.o\" a47=\"/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64/crtn.o\""
---
{
  "record_type": "EXECVE",
  "timestamp": 1655465398534,
  "id": 25618,
  "fields": {
    "a0": "/bin/sh",
    "a1": "-efu",
    "a10": "-plugin-opt=-pass-through=-lgcc",
    "a11": "-plugin-opt=-pass-through=-lgcc_s",
    "a12": "--build-id",
    "a13": "--no-add-needed",
    "a14": "--eh-frame-hdr",
    "a15": "--hash-style=gnu",
    "a16": "--as-needed",
    "a17": "-shared",
    "a18": "-X",
    "a19": "-EL",
    "a2": "/usr/bin/ld",
    "a20": "-maarch64linux",
    "a21": "-o",
    "a22": "ztest105133.so",
    "a23": "/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64/crti.o",
    "a24": "/usr/lib64/gcc/aarch64-alt-linux/8/crtbeginS.o",
    "a25": "-L/usr/lib64/gcc/aarch64-alt-linux/8",
    "a26": "-L/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64",
    "a27": "-L/lib/../lib64",
    "a28": "-L/usr/lib/../lib64",
    "a29": "-L/usr/lib64/gcc/aarch64-alt-linux/8/../../..",
    "a3": "-plugin",
    "a30": "-soname",
    "a31": "libz.so.1",
    "a32": "--version-script",
    "a33": "zlib.map",
    "a34": "ztest105133.o",
    "a35": "-lgcc",
    "a36": "--push-state",
    "a37": "--as-needed",
    "a38": "-lgcc_s",
    "a39": "--pop-state",
    "a4": "/usr/libexec/gcc/aarch64-alt-linux/8/liblto_plugin.so",
    "a40": "-lc",
    "a41": "-lgcc",
    "a42": "--push-state",
    "a43": "--as-needed",
    "a44": "-lgcc_s",
    "a45": "--pop-state",
    "a46": "/usr/lib64/gcc/aarch64-alt-linux/8/crtendS.o",
    "a47": "/usr/lib64/gcc/aarch64-alt-linux/8/../../../../lib64/crtn.o",
    "a5": "-plugin-opt=/usr/libexec/gcc/aarch64-alt-linux/8/lto-wrapper",
    "a6": "-plugin-opt=-fresolution=/usr/src/tmp/cchyHiZN.res",
    "a7": "-plugin-opt=-pass-through=-lgcc",
    "a8": "-plugin-opt=-pass-through=-lgcc_s",
    "a9": "-plugin-opt=-pass-through=-lc",
    "argc": "48"
  }
}
